runtime/connectors/blob/parquetreader.go

```diff
@@ -0,0 +1,146 @@
+package blob
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/apache/arrow/go/v11/arrow"
+	"github.com/apache/arrow/go/v11/arrow/array"
+	"github.com/apache/arrow/go/v11/arrow/memory"
+	"github.com/apache/arrow/go/v11/parquet"
+	"github.com/apache/arrow/go/v11/parquet/compress"
+	"github.com/apache/arrow/go/v11/parquet/file"
+	"github.com/apache/arrow/go/v11/parquet/pqarrow"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime/pkg/arrayutil"
+	"github.com/rilldata/rill/runtime/pkg/container"
+	"gocloud.dev/blob"
+)
+
+// number of rows of a column fetched in one call
+// keeping it high seems to improve latency at the cost of accuracy in size of fetched data as per policy
+const _batchSize = int64(1024)
+
+func downloadParquet(ctx context.Context, bucket *blob.Bucket, obj *blob.ListObject, option *extractOption, fw *os.File) error {
+	mem := memory.NewCheckedAllocator(memory.DefaultAllocator)
+	reader := NewBlobObjectReader(ctx, bucket, obj)
+
+	props := parquet.NewReaderProperties(mem)
+
+	pf, err := file.NewParquetReader(reader, file.WithReadProps(props))
+	if err != nil {
+		return err
+	}
+	defer pf.Close()
+
+	arrowReadProperties := pqarrow.ArrowReadProperties{BatchSize: _batchSize, Parallel: true}
+	// reader to convert parquet objects to arrow objects
+	fileReader, err := pqarrow.NewFileReader(pf, arrowReadProperties, mem)
+	if err != nil {
+		return err
+	}
+
+	numRowGroups := pf.NumRowGroups()
+	if numRowGroups == 0 {
+		return fmt.Errorf("invalid parquet")
+	}
+
+	records, err := estimateRecords(ctx, pf, fileReader, option)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		for _, rec := range records {
+			rec.Release()
+		}
+	}()
+
+	schema, err := fileReader.Schema()
+	if err != nil {
+		return err
+	}
+
+	table := array.NewTableFromRecords(schema, records)
+	defer table.Release()
+
+	// duck db requires root Repetitions to be required
+	// keeping compressions as uncompressed since file will be immediately consumed and deleted
+	wp := parquet.NewWriterProperties(
+		parquet.WithRootRepetition(parquet.Repetitions.Required),
+		parquet.WithCompression(compress.Codecs.Uncompressed),
```

Comment:
That's an interesting point. I did not implement any compression since we need to waste CPU cycles first in compressing then decompressing but agreed to your point of slow disk IO. Tough to make a decision without a benchmark. Will explore on this.

Reply:
We can keep the compression as Snappy which seems to be lightweight wrt cpu usage and achieves reasonable compression. Keeping the same compression as that of input file can be tricky since each column can have different compression. Please let me know your thoughts

Reply:
Snappy sounds good